Let the ListView handle item clicks and long-clicks

orac opened the pull request:
Fixes issue #13 and provides a starting point for fixing issue #6. The touch-handling code moves around a lot, but it's really pretty similar. onInterceptTouchEvent and onTouchEvent now just forward events (when appropriate) to a GestureDetector, and the callbacks from the GestureDetector do the real dragging and arranging work. The crucial change is that onInterceptTouchEvent returns false most of the time, which lets AbsListView do its item click detection the way it usually does.

There's definitely more to do after this: slide-to-remove isn't any better than it was, and should be easily improved now; also, with drag-to-remove turned on and no drag listener, it should still be possible to scroll the list (like in the recent apps list), but currently isn't.

orac committed:
Remove warnings in demos

Mostly unused imports, which I fixed by deleting. Also an unused member
in a nested class, which I also fixed by deleting.

orac committed:
Demo: avoid listing the launcher in the list of demos

It's just a question of filtering it out of the list returned by
PackageManager.

orac committed:
Add a demo for fling-to-remove

The new demo reuses all the code from BasicDSLV, but substitutes its own
layout (since you can only set removeMode in XML) and its own item
layout: I've moved the drag handle to the left side, to make it easier
to do the fling.

orac committed:
Don't start a drag immediately on touch down

When a touch-down action occurs, don't start a drag right away. Use a
GestureDetector to decide whether it is a tap or a drag, and if it isn't
a drag, let the ListView handle it in the usual way. This allows
onItemClick and onItemLongClick to work as normal, even when the whole
item is the drag handle.

orac committed:
Add a demo for handling item clicks when the whole item is draggable

The demo uses the standard OnItemClickListener and
OnItemLongClickListener to listen to item clicks, while the items are
also draggable.

bauerca commented:
A few comments. First, nice work! Second, on my phone, I notice that the floating View is sometimes empty; I think this is due to the changing of the text color (by ListView) when an item is selected. That is, the bitmap for the floating view is created from the list item in its selected state---we don't want this. Third, what if someone wants a drag to start immediately (i.e. as soon as a finger touches down on a `id:drag` subview)? For example, in the BasicDSLV demo, I expect the floating View to appear as soon as I touch the drag handle for an item; the selection/highlighting of the item immediately prior kinda throws me. Furthermore, I noticed that (also in the BasicDSLV demo) if I press on the drag handle of an item, but don't move it anywhere for long enough, no drag occurs at all.

I'd like to consider a longer-term solution that gives more control to the user. In particular, maybe DSLV shouldn't dictate to the developer how drags can be initiated (but it could certainly still provide shortcuts). I've recently thought that exposing a public `startDrag(int position, ViewBuilder builder)` method is the right thing to do. This also gets us closer to the customization available in the new Drag and Drop android framework, which developers expect.

I'll start a new branch (pushed to github) and merge your changes there (since I like to keep master free of bugs, albeit sometimes at the cost of fewer features). Maybe we can polish up your features and add them as convenience drag initiation enablers alongside a startDrag() framework. Sound good?

orac commented:
It's great that you're already thinking about things like that, as for my use case I want to let the user drag items between multiple DSLV instances, so integrating better with the Android DnD stuff (when running on a version that has it) is already something I'm planning to look at later this week.

Thanks for spotting the empty floating view problem. I don't see this on the device I've been testing on, but let me try it on a few more and see if I can reproduce it.

orac committed:
Improve long-press support

When starting a drag, set the dragged item to be not pressed before
taking a snapshot of its drawing cache.
This fixes the problem of the dragged item sometimes being highlighted
or invisible, depending on when you started the drag.

Add support for only starting a drag with a long-press. This allows you
to use the whole item layout as the drag handle but still let scrolling
work. Apologies for the name of the new attribute enumerator: using
"drag" makes aapt segfault because it has the same name as an existing
ID.

Start differentiating horizontal and vertical drags. This means that
when you have a remove handler but no drag handler, a vertical drag
still scrolls the list, but a horizontal drag can remove the item. This,
again, is especially useful when the whole item layout is the drag
handle, and makes the list work like Android's recent apps list. More
work is needed to distinguish horizontal and vertical drags when both
are enabled.

Add a demo for long-press-to-sort.

Add some text in the readme about (a) the new sort_mode attribute, and
(b) complications of having the drag handle be the whole list item.

orac commented:
The commit I've just added fixes the invisible text problem you noticed, and provides a solution for issue #6.

yincrash commented:
awesome!
